src/spartan/snark.rs

```diff
@@ -420,18 +421,14 @@ where
     };
 
     // compute evaluations of R1CS matrices
-    let multi_evaluate = |M_vec: &[&[(usize, usize, G::Scalar)]],
+    let multi_evaluate = |M_vec: &[&SparseMatrix<G::Scalar>],
                           r_x: &[G::Scalar],
                           r_y: &[G::Scalar]|
      -> Vec<G::Scalar> {
       let evaluate_with_table =
-        |M: &[(usize, usize, G::Scalar)], T_x: &[G::Scalar], T_y: &[G::Scalar]| -> G::Scalar {
-          (0..M.len())
-            .into_par_iter()
-            .map(|i| {
-              let (row, col, val) = M[i];
-              T_x[row] * T_y[col] * val
-            })
+        |M: &SparseMatrix<G::Scalar>, T_x: &[G::Scalar], T_y: &[G::Scalar]| -> G::Scalar {
+          M.iter()
+            .map(|(row, col, val)| T_x[row] * T_y[col] * val)
```

Comment:
In the matrix-vector multiplication above, we let the inherent shape of the matrix drive the operation, by iterating over the matrix in a row-by-row manner (and then indexing into the vector's array using the column index).

It seems that here, rather than using the slow iterator, we could do the same thing: proceed row-by-row on the native CSR shape, indec into T_x, T_y using row & column indices.